safehttp/plugins/xsrf/xsrf.go

```diff
@@ -18,146 +18,229 @@
-// Forgery attack.
-//
-// On first user visit through a state preserving request (GET, HEAD or
-// OPTIONS), a nonce-based cookie will be set in the response as a way to
-// distinguish between users and prevent pre-login XSRF attacks. The cookie will
-// be used in the token generation and verification algorithm and is expected to
-// be present in all subsequent incoming requests.
-//
-// For every authorized request, the interceptor will also generate a
-// cryptographically-safe XSRF token using the appKey, the cookie and the path
-// visited. This can be later extracted using Token and should be injected as a
-// hidden input field in HTML forms.
-//
-// In case of state changing requests (all except GET, HEAD and OPTIONS), the
-// interceptor checks for the presence of the XSRF token in the request body
-// (expected to have been injected) and validates it.
 func (it *Interceptor) Before(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, _ safehttp.InterceptorConfig) safehttp.Result {
-	needsValidation := !statePreservingMethods[r.Method()]
-	cookieID, err := r.Cookie(cookieIDKey)
+	if statePreservingMethods[r.Method()] {
+		return safehttp.NotWritten()
+	}
+	tok, userID, actionID, code := it.c.Retrieve(r)
+	if code != safehttp.StatusOK {
+		return w.WriteError(code)
+	}
+	code = it.c.Validate(tok, userID, actionID)
+	if code != safehttp.StatusOK {
+		return w.WriteError(code)
+	}
+	return safehttp.NotWritten()
+}
+
+func (it *Interceptor) Commit(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, resp safehttp.Response, _ safehttp.InterceptorConfig) safehttp.Result {
+	err := it.i.Inject(resp, w, r)
 	if err != nil {
-		if needsValidation {
-			return w.WriteError(safehttp.StatusForbidden)
-		}
-		cookieID, err = addCookieID(w)
-		if err != nil {
-			// An error is returned when the plugin fails to Set the Set-Cookie
-			// header in the response writer as this is a server misconfiguration.
-			return w.WriteError(safehttp.StatusInternalServerError)
-		}
+		return w.WriteError(safehttp.StatusInternalServerError)
+	}
+	return safehttp.Result{}
+}
+
+// Checker decides whether a safehttp.IncomingRequest should be allowed to pass
+// or it's part of a Cross-Site Request Forgery (XSRF) attack.
+type Checker interface {
+	// Retrieve returns the information necessary to verify the validity of
+	// a XSRF token.
+	Retrieve(r *safehttp.IncomingRequest) (token, userID, actionID string, code safehttp.StatusCode)
+	// Validate validates the token contained in the request, optionally using
+	// the user session and action the user is taking.
+	Validate(token string, userID string, actionID string) safehttp.StatusCode
+}
+
+type Injector interface {
+	// Inject adds the protection necessary to the safehttp.Response and
+	// safehttp.ResponseWriter so that subsequent request can be deemed safe.
+	Inject(resp safehttp.Response, w *safehttp.ResponseWriter, r *safehttp.IncomingRequest) error
```

Comment:
Nite: for sake of coherence let's preserve the order of the Commit phase: w, r, resp